pkg/addon/agent/agent.go

```diff
@@ -0,0 +1,276 @@
+					{
+						SignerName: ProxyAgentSignerName,
+						Subject: addonv1alpha1.Subject{
+							User: common.SubjectUserClusterProxyAgent,
+							Groups: []string{
+								common.SubjectGroupClusterProxy,
+							},
+						},
+					},
+				}
+			},
+			CSRApproveCheck: func(cluster *clusterv1.ManagedCluster, addon *addonv1alpha1.ManagedClusterAddOn, csr *csrv1.CertificateSigningRequest) bool {
+				return cluster.Spec.HubAcceptsClient
+			},
+			PermissionConfig: func(cluster *clusterv1.ManagedCluster, addon *addonv1alpha1.ManagedClusterAddOn) error {
+				return nil
+			},
+			CSRSign: func(csr *csrv1.CertificateSigningRequest) []byte {
+				if csr.Spec.SignerName != ProxyAgentSignerName {
+					return nil
+				}
+				b, _ := pem.Decode(csr.Spec.Request)
+				parsed, err := x509.ParseCertificateRequest(b.Bytes)
+				if err != nil {
+					return nil
+				}
+				validity := time.Hour * 24 * 180
+				caCert := p.selfSigner.CA().Config.Certs[0]
+				tmpl := &x509.Certificate{
+					SerialNumber:       caCert.SerialNumber,
+					Subject:            parsed.Subject,
+					DNSNames:           parsed.DNSNames,
+					IPAddresses:        parsed.IPAddresses,
+					EmailAddresses:     parsed.EmailAddresses,
+					URIs:               parsed.URIs,
+					PublicKeyAlgorithm: parsed.PublicKeyAlgorithm,
+					PublicKey:          parsed.PublicKey,
+					Extensions:         parsed.Extensions,
+					ExtraExtensions:    parsed.ExtraExtensions,
+					IsCA:               false,
+					KeyUsage:           x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+					ExtKeyUsage: []x509.ExtKeyUsage{
+						x509.ExtKeyUsageServerAuth,
+						x509.ExtKeyUsageClientAuth,
+					},
+				}
+				now := time.Now()
+				tmpl.NotBefore = now
+				tmpl.NotAfter = now.Add(validity)
+
+				rsaKey := p.selfSigner.CA().Config.Key.(*rsa.PrivateKey)
+				der, err := x509.CreateCertificate(
+					rand.Reader,
+					tmpl,
+					p.selfSigner.CA().Config.Certs[0],
+					parsed.PublicKey,
+					rsaKey)
+				if err != nil {
+					return nil
+				}
+				return pem.EncodeToMemory(&pem.Block{
+					Type:  "CERTIFICATE",
+					Bytes: der,
+				})
+			},
+		},
+	}
+}
+
+const (
+	ApiserverNetworkProxyLabelAddon     = "open-cluster-management.io/addon"
+	ApiserverNetworkProxyLabelComponent = "open-cluster-management.io/component"
+
+	AgentSecretName   = "cluster-proxy-open-cluster-management.io-proxy-agent-signer-client-cert"
+	AgentCASecretName = "cluster-proxy-ca"
+)
+
+func newAgentDeployment(clusterName, targetNamespace string, proxyConfig *proxyv1alpha1.ManagedProxyConfiguration, loadBalancerEndpoint string) *appsv1.Deployment {
+	serviceEntryPoint := proxyConfig.Spec.ProxyServer.InClusterServiceName + "." + proxyConfig.Spec.ProxyServer.Namespace
+	if len(proxyConfig.Spec.ProxyAgent.ProxyServerHost) > 0 {
+		serviceEntryPoint = proxyConfig.Spec.ProxyAgent.ProxyServerHost
+	} else if len(loadBalancerEndpoint) > 0 {
+		serviceEntryPoint = loadBalancerEndpoint
+	}
```

Comment:
we may need handle the case that if there is no `proxyConfig.Spec.ProxyAgent.ProxyServerHost` and the service is not lb,

Reply:
and could we get the `loadBalancerEndpoint` at here, I think it may be more clear, and it's better to record some log here

Reply:
clarified the logic of endpoint resolving